File: 1-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {

        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];

            if(map.containsKey(diff)) {
                return new int[] {map.get(diff), i};
            } else {
                map.put(nums[i], i);
            }
        }

        return new int[0];
    }
}

// Notes:
// Two main approaches

// 1 ) Sort nums
// then use two pointers <left, right> and inc/dec till you get target

// 2 ) HashMap
// Iterate through arr, check map for diff (target - nums[i]), if diff doesnt exist add <nums[i], i> to map and continue
// This works because we can check the required difference value/index, and we only add the current value/index to the hashmap
// after we check, so we dont have to worry about duplicates.

File: 1-two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        num_set = {}

        for index, num in enumerate(nums):
            diff = target - num

            if diff in num_set:
                #NeetCode requires it to be sorted
                #more eff w/an if else but, thisis fine
                return sorted([index, num_set[diff]])

            num_set[num] = index

        #default empty
        return []